client/state/data-layer/wpcom-http/test/index.js

```diff
@@ -51,11 +67,11 @@ describe( '#queueRequest', () => {
 
 		http( { dispatch }, getMe );
 
-		setTimeout( () => {
+		afterNockDone( nock, () => {
 			expect( dispatch ).to.have.been.calledOnce;
 			expect( dispatch ).to.have.been.calledWith( extendAction( succeeder, successMeta( data ) ) );
 			done();
-		}, 10 );
+		} );
```

Comment:
note that `sinon` has `fakeTimers` and `Jest` has `jest.runAllTimers()`
https://facebook.github.io/jest/docs/en/timer-mocks.html

so while I'm embarrassed that I introduced this test race, I would love
to see if we can't fix it by eliminating the asynchronicity instead of just
making our tests run slower to patch over it.

with a little experimentation I ran into some issues actually making the
timers advance (Sinon seemed to work better than Jest with regards
to timers) but I believe that it should in the end be simple to do.

Reply:
Using `jest.useFakeTimers()` and `jest.runAllTimers()` doesn't work when I try to use them to force `nock` to execute the mocked response callbacks. Maybe I could make it work after some research, but I'm afraid I would have to know too many `nock` implementation details to be comfortable with that.

There are no `setTimeout`s in the **tested** code. They are inside the `nock` library (and I'm not even sure about that). I'd rather keep the current approach, where we simply wait without any clever tricks.

Are you sure the tests are really running slower after this patch? Why would that be so? How big do you think the slowdown is, if there is any?

Reply:
> Using jest.useFakeTimers() and jest.runAllTimers() doesn't work when I try to use them to force nock to execute the mocked response callbacks

This is what I feared. It's a big issue with Jest too because `Sinon` provided a very nice interface while it appears like the Jest authors never even anticipated peoples' use-cases for this stuff.

> There are no setTimeouts in the tested code. They are inside the nock library (and I'm not even sure about that).

The testing library just changes the global `setTimeout` and `setInterval` functions so there's no magic happening or need to dive into `nock`

> I'd rather keep the current approach, where we simply wait without any clever tricks.

Well the point was that adding timeouts inside of tests is the tricky bits and forcing a specific time behavior through mocking is the reliable non-clever way of doing it. If this is too much of a problem though it's okay to keep the timeouts in there.

> Are you sure the tests are really running slower after this patch?

No but fairly confident unless there is other mocking going on not visible here.

> Why would that be so? 

Because the test code itself literally sets a timeout and waits to resolve.

> How big do you think the slowdown is, if there is any?

In this case it should jump from 20ms to up to 200ms of testing delay that isn't needed (but probably in most cases won't add any, or add an extra 20ms for a single retry). For some reason I can't remember, `nock` isn't forwarding its payload until after a real delay even though it could. It's probably just some default. We might be able to try settings its delay parameter to `0` with `.delayBody( 0 )` but I don't know if that would work.

---

again, if this is too much work just merge it in, but instead of replacing something that breaks some of the time with something that breaks less of the time I'd rather just replace it with something that doesn't break :wink:

Reply:
> The testing library just changes the global setTimeout and setInterval functions so there's no magic happening or need to dive into nock

`jest.useFakeTimeouts` doesn't work, however. So I needed to dive into `nock` and found out that it uses `setImmediate` not from the global object, but imported from the `timers` module! That module has the real version instead of the mock.

Also, I found out that `nock.isDone()` means something else than I thought. The return value becomes `false` at the moment when `nock` is finished working with the mocked request interceptor, but **before** it calls `process.nextTick(respond)`. So again, our asserts can still sometimes run before the response events are fired and `dispatch` called.

I ended up rewriting the tests to wait until a `dispatch` mock is called, then verifying its arguments and finally calling the test's `done` callback. Please check it out.